cluster-autoscaler/proposals/circumvent-tag-limit-aws.md

```diff
@@ -0,0 +1,142 @@
+      "version": "*string*"
+   }
+}
+```
+
+[Current EKS interface]: https://github.com/aws/aws-sdk-go/blob/82e096143fdfb8f52fbeb4ef78d400eea6381ccd/service/eks/eksiface/interface.go#L96
+[DescribeNodegroup API]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html
+[here]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html#API_DescribeNodegroup_ResponseSyntax
+
+## Implementation Proposal
+
+1. In the autoscaling loop, Cluster Autoscaler finds no nodes, so it will build the nodeTemplate from other information.
+1. Cluster Autoscaler will first get some common allocatable resource values. ([allocatable resources code][])
+1. Cluster Autoscaler will then check the ASG tags for the EKS ManagedNodegroups tags. ([ASG tags code][])
+1. If Cluster Autoscaler finds the EKS ManagedNodegroups tags, it will call methods on the ManagedNodegroupCache to get labels, taints, and some other values.
+   1. If the cache has data for the given ManagedNodegroup, it will return that data.
+   1. If the data’s TTL has expired, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+   1. If there is no data cached for the ManagedNodegroup, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+1. Then Cluster Autoscaler will pull information from the ASG tags like it already does. If duplicate keys are found, the values from the ASG tags are kept.
+1. Cluster Autoscaler will then continue as usual.
+
+[allocatable resources code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L366
+[ASG tags code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L371
+
+## Testing
+
+- [TC1] Test creating a DescribeNodegroup request
+- [TC2] Test pulling values from cache when data exists for the ManagedNodegroups and TTL has not expired
+- [TC3] Test pulling values from cache and calling DescribeNodegroup API if no data exists
+- [TC4] Test pulling values from cache and calling DescribeNodegroup API if TTL has expired
+- [TC5] Test missing permissions and falling back to just ASG tags
+- [TC6] Test duplicate label keys - selecting the value from ASG tags response
+
+## Risks
+
+Latency and throttling:
+
+By default, Cluster Autoscaler runs every 10 seconds. Our best practices documentation notes that this short interval can cause throttling because Cluster Autoscaler already makes AWS API calls during each loop. Our documentation recommends that customers increase the interval, so adding this API shouldn’t cause latency problems for customers. ([EKS Best Practices][]) Also, the API call will only happen for EKS ManagedNodegroups. If this increase in latency is too much for even one run of the loop, we will look into moving the API calls into parallel goroutines. 
```

Comment:
i'm just curious if there is precedent for using parallel goroutines in cluster-autoscaler to do similar types of API calls?

Reply:
Thank you for taking a look at my proposal.
There's no precedent for using parallel goroutines to do similar types of API calls in the AWS cloud-provider code. I'm less familiar with the rest of the code, but I don't think there's a precedent in it either.

Reply:
There is a precedent in GCE provider: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/gce/cache.go#L277.